lib/org/chromium/webidl/Canonicalizer.js

```diff
@@ -0,0 +1,129 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+'use strict';
+
+foam.CLASS({
+  package: 'org.chromium.webidl',
+  name: 'Canonicalizer',
+
+  documentation: `This component expects to receive AST nodes from various
+    sources. Canonical versions of interface definitions will be created
+    from AST nodes of the same source.
+    Please note that this step cannot be distributed. All computations
+    must be done through one canonicalizer instances, or you will
+    get fragmented results.`,
+
+  requires: [
+    'org.chromium.webidl.ast.Enum',
+    'org.chromium.webidl.ast.Typedef',
+    'org.chromium.webidl.PipelineMessage',
+  ],
+
+  properties: [
+    {
+      class: 'Function',
+      documentation: `Function is called after waitTime seconds has passed
+        since the last call to addFragment.`,
+      name: 'onDone',
+      required: true,
+    },
+    {
+      class: 'Int',
+      documentation: `The amount of time to wait in seconds after the last
+        addFragment request, before onDone() is called with the canonical
+        IDL files.`,
+      name: 'waitTime',
+    },
+    {
+      documentation: `A map of maps containing canonical IDL files. The
+        key of the map corresponds to the renderer (source) from which the
+        files were extract. For example, rendererMap['blink'] is a map that
+        contains the canonical IDL files from the Blink repo.`,
+      name: 'rendererMap_',
+      factory: function() { return {}; },
+    },
+    {
+      documentation: `A map of Timers that is used to determine whether waitTime
+        has passed since the last call to addFragment.`,
+      name: 'timerMap_',
+      factory: function() { return {}; },
+    },
+  ],
+
+  methods: [
+    {
+      documentation: `Takes a given AST and extracts the definition members
+        from the AST. The members are then placed into the map corresponding to
+        the given source and interface.`,
+      name: 'addFragment',
+      code: function(source, ast, metadata) {
+        // Determine if Map exists for this source. Create one if non-existent.
+        var map = this.rendererMap_[source];
+        if (!map) {
+          map = {};
+          this.rendererMap_[source] = map;
+        }
+
+        var def = ast.definition;
+        // Ignore definition if Enum or Typedef.
```

Comment:
Is this the right thing to do? Should this component store them? Warn or throw on name reuse? Leaving a `TODO` might be sufficient here, but I'm not convinced that dropping these is the right choice.